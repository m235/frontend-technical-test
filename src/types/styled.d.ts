import 'styled-components'

declare module 'styled-components' {
  export interface DefaultTheme {
    palette: {
      primary: string
      secondary: string
      backgrounds: {
        light: string
        contrasted: string
      }
      grays: {
        lighter: string
        ligth: string
        normal: string
        dark: string
        darker: string
      }
      texts: {
        dark: string
        light: string
      }
    }
    spacing: {
      s050: string
      s100: string
      s200: string
      s250: string
      s300: string
      s400: string
    }
    elevation: {
      z0: string
      z1: string
    }
    breakpoint: {
      tablet: string
      desktop: string
    }
  }
}
