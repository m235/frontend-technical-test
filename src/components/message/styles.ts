import styled, { css } from 'styled-components'

const mineStyle = css`
  justify-content: flex-end;
  align-items: flex-end;
  border-radius: 8px;
  
  background-color: ${({ theme }) => theme.palette.primary};
  color: #ffffff;
`

const theirsStyle = css`
  justify-content: flex-start;
  align-items: flex-start;
  
  border-radius: 8px;
  
  background-color: ${({ theme }) => theme.palette.backgrounds.light};
  color: rgb(26, 26, 26);
`

export const MessageContainer = styled.div<{ isMine: boolean }>`
  display: flex;
  padding: ${({ theme }) => theme.spacing.s200};
  
  font-size: 14px;
  ${({ isMine }) => (isMine ? mineStyle : theirsStyle)};
  word-break: break-word;
`
