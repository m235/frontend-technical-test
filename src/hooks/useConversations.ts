import { AxiosError } from 'axios'
import { useQuery } from 'react-query'

import { ConversationsApi, getConversations } from '@/services/conversations'
import type { Conversation } from '@/types/conversation'

const useConversations = (userId?: number) => {
  return useQuery<Conversation[], AxiosError>([ConversationsApi.CONVERSATIONS, userId], () => getConversations(userId), {
    enabled: !!userId,
    useErrorBoundary: true,
  })
}

export default useConversations
