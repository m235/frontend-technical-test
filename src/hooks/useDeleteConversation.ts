import { AxiosError } from 'axios'
import { useMutation } from 'react-query'
import { UseMutationOptions } from 'react-query/types/react/types'

import type { Conversation } from '@/types/conversation'
import { deleteConversation } from '@/services/conversations'

type Options = UseMutationOptions<Conversation, AxiosError, number>

const useDeleteConversation = (conversationId: number, options?: Options) => {
  return useMutation<Conversation, AxiosError, number>(deleteConversation, options)
}

export default useDeleteConversation
