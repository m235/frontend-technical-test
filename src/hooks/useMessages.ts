import type { AxiosError } from 'axios'
import { useQuery } from 'react-query'

import { getMessages, MessagesApi } from '@/services/messages'
import type { Message } from '@/types/message'

const useMessages = (conversationId: number) => {
  return useQuery<Message[], AxiosError>([MessagesApi.MESSAGES, conversationId], () => getMessages(conversationId), {
    enabled: !!conversationId,
    useErrorBoundary: true,
  })
}

export default useMessages
