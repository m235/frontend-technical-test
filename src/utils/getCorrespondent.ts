import type { Conversation } from '@/types/conversation'

export const getCorrespondent = (userId: number, conversation: Conversation) => {
  switch (userId) {
    case conversation.senderId:
      return conversation.recipientNickname
    case conversation.recipientId:
      return conversation.senderNickname
    default:
      throw new Error('You are nor the recipient nor the sender')
  }
}
