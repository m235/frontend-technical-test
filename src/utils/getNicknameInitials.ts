export const getNicknameInitials = (nickname: string) =>
  nickname
    .split(' ')
    .map((string) => string.trim().charAt(0))
    .slice(0, 2)
    .join('')
