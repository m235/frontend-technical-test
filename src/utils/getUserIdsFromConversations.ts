import type { Conversation } from '@/types/conversation'

export const getUserIdsFromConversations = (conversations?: Conversation[]) =>
  Array.from<number>(new Set(conversations?.reduce((acc, curr) => [...acc, curr.recipientId, curr.senderId], []) ?? []))
