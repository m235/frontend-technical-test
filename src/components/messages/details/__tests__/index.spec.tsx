import useAuth from '@/hooks/useAuth'
import useMessages from '@/hooks/useMessages'
import { render } from '@/tests/utils'
import type { Conversation } from '@/types/conversation'
import { Message } from '@/types/message'
import MessagesDetails from '@/components/messages/details'
import useConversations from '@/hooks/useConversations'
import useDeleteConversation from '@/hooks/useDeleteConversation'
import { fireEvent } from '@testing-library/dom'

const mockedUseAuth = useAuth as jest.Mock<any>
const mockedUseConversations = useConversations as jest.Mock<any>
const mockedUseDeleteConversation = useDeleteConversation as jest.Mock<any>
const mockedUseMessages = useMessages as jest.Mock<any>
jest.mock('@/hooks/useAuth')
jest.mock('@/hooks/useConversations')
jest.mock('@/hooks/useMessages')
jest.mock('@/hooks/useDeleteConversation')

describe('components:conversations:details', () => {
  const conversations: Conversation[] = [
    {
      id: 1,
      recipientNickname: 'Bob',
      recipientId: 2,
      senderNickname: 'John Travolta',
      senderId: 1,
    },
    {
      id: 2,
      recipientNickname: 'John Travolta',
      recipientId: 1,
      senderNickname: 'Johnny Hallyday',
      senderId: 3,
    },
  ]
  const messages: Message[] = [
    { id: 1, timestamp: new Date().getTime(), authorId: 1, conversationId: 1, body: 'Foo' },
    { id: 2, timestamp: new Date().getTime(), authorId: 2, conversationId: 1, body: 'Bar' },
  ]

  beforeEach(() => {
    mockedUseAuth.mockImplementation(() => ({ userId: 1, user: [{ id: 1, nickname: 'John Travolta', token: 'x' }] }))
    mockedUseConversations.mockImplementation(() => ({ isLoading: false, data: conversations }))
    mockedUseMessages.mockImplementation(() => ({ isLoading: true }))
    mockedUseDeleteConversation.mockImplementation(() => ({ isLoading: true, mutate: () => null }))
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render default correctly', () => {
    const { container } = render(<MessagesDetails conversationId={1} />)

    expect(container).toMatchSnapshot()
  })
  it('should render loading messages', () => {
    const { getByTestId } = render(<MessagesDetails conversationId={1} />)

    expect(getByTestId('text-skeleton')).toBeInTheDocument()
  })
  it('should not render the delete conversation button', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: [] }))
    const { queryByTestId } = render(<MessagesDetails conversationId={2} />)

    expect(queryByTestId('delete-conversation')).not.toBeInTheDocument()
  })
  it('should render the delete conversation button', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: [] }))
    const { getByTestId } = render(<MessagesDetails conversationId={1} />)

    expect(getByTestId('delete-conversation')).toBeInTheDocument()
  })
  it('should render the last message date', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: messages }))
    const { queryByTestId } = render(<MessagesDetails conversationId={2} />)

    expect(queryByTestId('last-message-date')).toBeInTheDocument()
  })
  it('should set the button has disabled', () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: messages }))
    const { getByTestId } = render(<MessagesDetails conversationId={1} />)

    expect(getByTestId('delete-conversation')).toBeDisabled()
  })
  it('should call the mutation function with conversation id', () => {
    const deleteFn = jest.fn()
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: messages }))
    mockedUseDeleteConversation.mockImplementation(() => ({ isLoading: false, mutate: deleteFn }))
    const { getByTestId } = render(<MessagesDetails conversationId={1} />)

    fireEvent.click(getByTestId('delete-conversation'))
    expect(deleteFn).toHaveBeenCalledTimes(1)
    expect(deleteFn).toHaveBeenCalledWith(1)
  })
})
