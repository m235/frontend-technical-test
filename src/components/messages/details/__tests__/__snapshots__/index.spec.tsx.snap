// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:conversations:details should render default correctly 1`] = `
<div>
  <aside
    class="sc-eCYdqJ cGbBXZ"
  >
    <div
      class="sc-gKXOVf cFizLK"
    >
      <div
        class="sc-gsnTZi eUelFf sc-hKMtZM iEftAl"
      >
        B
      </div>
      <h2
        class="sc-jSMfEi eLjiwY"
      >
        Conversation with 
        <span>
          Bob
        </span>
      </h2>
      <span
        class="sc-bczRLJ gLjPvE"
        data-testid="text-skeleton"
      />
    </div>
    <button
      class="sc-dkzDqf exBfzp sc-ftvSup iAEcyI"
      data-testid="delete-conversation"
      disabled=""
    >
      <i
        class="fa-solid fa-trash-can"
      />
       Delete this conversation
    </button>
  </aside>
</div>
`;
